first attempt to fix molecule

pwalczysko opened the pull request:
According to the changes in https://github.com/ome/ansible-role-ssl-certificate/pull/17/files, this is a first attempt to fix the molecule build in this repo.


cc @sbesson @jburel 


Status:

1. The framework in this repo was adjusted according to https://github.com/IDR/deployment/pull/397/files as far as I understood
2. Confusing points:
    - Is the python and other versions to be overtaken from https://github.com/ome/ansible-role-ssl-certificate/commit/92e5050beb61349ade69dbcf8632b662182f6e6b or https://github.com/IDR/deployment/pull/397/files in respect to the work to be done in this repo ?
    - The problem is that I tried both sets of versions as listed in the previous bullet point and failed to get the build running
    - Is it important that this repo does not have any `default` subfolder of the folder `molecule` ? If yes, should a new subfolder structure be created in this repo ? 
    - It seems that the structure in this repo is more complex than in the https://github.com/ome/ansible-role-ssl-certificate - in this repo, we have around 10 different ``molecule.yml`` files, not just one. Also, in this repo, there is a subfolder called ``k8s``  - probably for Kubernetes https://github.com/ome/prod-playbooks/tree/master/k8s ? Is that of any significance ?

pwalczysko commented:
```
  molecule test -s "bootstrap"
  shell: /usr/bin/bash -e {0}
  env:
    pythonLocation: /opt/hostedtoolcache/Python/3.9.16/x64
    PKG_CONFIG_PATH: /opt/hostedtoolcache/Python/3.9.16/x64/lib/pkgconfig
    Python_ROOT_DIR: /opt/hostedtoolcache/Python/3.9.16/x64
    Python2_ROOT_DIR: /opt/hostedtoolcache/Python/3.9.16/x64
    Python3_ROOT_DIR: /opt/hostedtoolcache/Python/3.9.16/x64
    LD_LIBRARY_PATH: /opt/hostedtoolcache/Python/3.9.16/x64/lib
CRITICAL Failed to validate /home/runner/work/prod-playbooks/prod-playbooks/molecule/bootstrap/molecule.yml

["Additional properties are not allowed ('lint' was unexpected)"]
Error: Process completed with exit code 1.
```

sorry, I am stuck on this ^^^ - does not seem to resolve whatever I do, no idea who is even running that test...
Any ideas @sbesson please ?

sbesson commented:
> sorry, I am stuck on this ^^^ - does not seem to resolve whatever I do, no idea who is even running that test...
Any ideas @sbesson please ?

From https://github.com/ome/ansible-role-ssl-certificate/pull/17/commits/92e5050beb61349ade69dbcf8632b662182f6e6b, you also need to remove 
```
  lint:
    name: flake8
 ```
from the `verifier` phase at the end of each Molecule configuration

pwalczysko commented:
@sbesson Thaks that helped

But the new error is - this time does not even mention what it causes...

```
INFO     Running bootstrap > lint
Molecule bootstrap > lint
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.9.16/x64/bin/molecule", line 8, in <module>
    sys.exit(main())
  File "/opt/hostedtoolcache/Python/3.9.16/x64/lib/python3.9/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/opt/hostedtoolcache/Python/3.9.16/x64/lib/python3.9/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/opt/hostedtoolcache/Python/3.9.16/x64/lib/python3.9/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/opt/hostedtoolcache/Python/3.9.16/x64/lib/python3.9/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/opt/hostedtoolcache/Python/3.9.16/x64/lib/python3.9/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/opt/hostedtoolcache/Python/3.9.16/x64/lib/python3.9/site-packages/click/decorators.py", line 26, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/opt/hostedtoolcache/Python/3.9.16/x64/lib/python3.9/site-packages/molecule/command/test.py", line 176, in test
    base.execute_cmdline_scenarios(scenario_name, args, command_args, ansible_args)
  File "/opt/hostedtoolcache/Python/3.9.16/x64/lib/python3.9/site-packages/molecule/command/base.py", line 121, in execute_cmdline_scenarios
    execute_scenario(scenario)
  File "/opt/hostedtoolcache/Python/3.9.16/x64/lib/python3.9/site-packages/molecule/command/base.py", line 165, in execute_scenario
    execute_subcommand(scenario.config, action)
  File "/opt/hostedtoolcache/Python/3.9.16/x64/lib/python3.9/site-packages/molecule/command/base.py", line 154, in execute_subcommand
    return command(config).execute(args)
  File "/opt/hostedtoolcache/Python/3.9.16/x64/lib/python3.9/site-packages/molecule/logger.py", line 188, in wrapper
    rt = func(*args, **kwargs)
  File "/opt/hostedtoolcache/Python/3.9.16/x64/lib/python3.9/site-packages/molecule/logger.py", line 99, in wrapper
    return func(*args, **kwargs)
  File "/opt/hostedtoolcache/Python/3.9.16/x64/lib/python3.9/site-packages/molecule/command/lint.py", line 97, in execute
    result = util.run_command(cmd, env=self.env, echo=True)
  File "/opt/hostedtoolcache/Python/3.9.16/x64/lib/python3.9/site-packages/molecule/util.py", line 150, in run_command
    result = app.runtime.exec(
AttributeError: 'Runtime' object has no attribute 'exec'
Error: Process completed with exit code 1.
```

pwalczysko commented:
@sbesson actually, the error is cited in https://github.com/ansible-community/molecule/issues/3903 - but not sure how to solve...

pwalczysko commented:
And the new error is

either

```
CRITICAL Couldn't parse task at omero/training-server/playbook.yml:400 (couldn't resolve module/action 'iptables_raw_25'. This often indicates a misspelling, missing collection, or incorrect module path.)
{ 'become': True,
  'iptables_raw_25': { '__file__': 'omero/training-server/playbook.yml',
                       '__line__': 403,
                       'name': 'docker_outreach_rules',
                       'rules': '-A FORWARD -i docker0 -o {{ external_nic }} '
                                '-j ACCEPT\n'
                                '-A FORWARD -i {{ external_nic }} -o docker0 '
                                '-j ACCEPT\n',
                       'state': 'present'},
  'name': 'Iptables Docker forward rules'}
```

but then below there is another error

```
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.8.16/x64/bin/molecule", line 8, in <module>
    sys.exit(main())
  File "/opt/hostedtoolcache/Python/3.8.16/x64/lib/python3.8/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/opt/hostedtoolcache/Python/3.8.16/x64/lib/python3.8/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/opt/hostedtoolcache/Python/3.8.16/x64/lib/python3.8/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/opt/hostedtoolcache/Python/3.8.16/x64/lib/python3.8/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/opt/hostedtoolcache/Python/3.8.16/x64/lib/python3.8/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/opt/hostedtoolcache/Python/3.8.16/x64/lib/python3.8/site-packages/click/decorators.py", line 26, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/opt/hostedtoolcache/Python/3.8.16/x64/lib/python3.8/site-packages/molecule/command/test.py", line 176, in test
    base.execute_cmdline_scenarios(scenario_name, args, command_args, ansible_args)
  File "/opt/hostedtoolcache/Python/3.8.16/x64/lib/python3.8/site-packages/molecule/command/base.py", line 132, in execute_cmdline_scenarios
    execute_subcommand(scenario.config, "destroy")
  File "/opt/hostedtoolcache/Python/3.8.16/x64/lib/python3.8/site-packages/molecule/command/base.py", line 154, in execute_subcommand
    return command(config).execute(args)
  File "/opt/hostedtoolcache/Python/3.8.16/x64/lib/python3.8/site-packages/molecule/logger.py", line 188, in wrapper
    rt = func(*args, **kwargs)
  File "/opt/hostedtoolcache/Python/3.8.16/x64/lib/python3.8/site-packages/molecule/logger.py", line 99, in wrapper
    return func(*args, **kwargs)
  File "/opt/hostedtoolcache/Python/3.8.16/x64/lib/python3.8/site-packages/molecule/command/destroy.py", line 107, in execute
    self._config.provisioner.destroy()
  File "/opt/hostedtoolcache/Python/3.8.16/x64/lib/python3.8/site-packages/molecule/provisioner/ansible.py", line 744, in destroy
    pb.execute()
  File "/opt/hostedtoolcache/Python/3.8.16/x64/lib/python3.8/site-packages/molecule/provisioner/ansible_playbook.py", line 117, in execute
    self._config.driver.sanity_checks()
  File "/opt/hostedtoolcache/Python/3.8.16/x64/lib/python3.8/site-packages/molecule_plugins/docker/driver.py", line 245, in sanity_checks
    docker_client = docker.from_env()
  File "/opt/hostedtoolcache/Python/3.8.16/x64/lib/python3.8/site-packages/docker/client.py", line 96, in from_env
    return cls(
  File "/opt/hostedtoolcache/Python/3.8.16/x64/lib/python3.8/site-packages/docker/client.py", line 45, in __init__
    self.api = APIClient(*args, **kwargs)
  File "/opt/hostedtoolcache/Python/3.8.16/x64/lib/python3.8/site-packages/docker/api/client.py", line 197, in __init__
    self._version = self._retrieve_server_version()
  File "/opt/hostedtoolcache/Python/3.8.16/x64/lib/python3.8/site-packages/docker/api/client.py", line 221, in _retrieve_server_version
    raise DockerException(
docker.errors.DockerException: Error while fetching server API version: request() got an unexpected keyword argument 'chunked'
```

pwalczysko commented:
@sbesson updated the header of this PR with some summary - no idea where to go from here i am afraid

pwalczysko commented:
Maybe also here the molecule is failing on syntactic errors ? There are plenty of reports like:

```
./release/public-images.yml
  Warning: 7:16 [truthy] truthy value should be one of [false, true]
  Error: 9:81 [line-length] line too long (81 > 80 characters)
  Warning: 14:16 [truthy] truthy value should be one of [false, true]
  Error: 18:49 [new-line-at-end-of-file] no new line character at the end of file
```

still a mystery to me which error is fatal and which is not

pwalczysko commented:
@sbesson as discussed today, fixed all the linting errors in all the playbooks reported (e.g. too long lines, indentation, block endings).

But still all the jobs are failing with 5 identical errors. These errors can be divided into 3 classes

- https://github.com/ome/prod-playbooks/pull/370/files#diff-89606df895216c428c9e3283542d38d46ff66653c253bcdbedf6b78c27f9d724
it is complaining about a line which I did not change - ``A playbook must be a list of plays, got a <class 'ansible.parsing.yaml.objects.AnsibleMapping'> instead``, the line is

- hosts: "{{ omero_restore_host }}" in https://github.com/ome/prod-playbooks/pull/370/files#diff-4445b3b683b2179dd7ce77967960394a08fe4a7cfc8c378c68a1bbb6f57e9976
variable not defined - again, did not change this line

- include: letsencrypt.yml (repeats in 2 other cases as well, other ``include`` lines are not passing in other plbooks)
does not like "include" lines, claiming this is not valid param for Play in 3 playbooks, e.g. last lines in https://github.com/ome/prod-playbooks/pull/370/files#diff-5b17bd29c054edd015ec7fe00c8bca4d08d645766c9684d7cc31a6f654c842c8
again, did not change these lines

It seems to me that all 5 errors have something to do with the passing of variables and dependencies.

Is it okay to ignore them ?
Or is it some change in the overall architecture which I caused by overtaking your changes from the "ansible-compat<4" repo ?

Thanks

sbesson commented:
I am unsure about the first exception but the second one can probably be fixed by declaring a default value `hosts: "{{ omero_restore_host | default("localhost") }}"` and the first one should be addressed by changing `include` into `import_playbook` which is the new accepted form for sub-playbooks - see https://github.com/IDR/deployment/commit/2203a565b2dbae07086f876db7d735fb4187a045

pwalczysko commented on .config/ansible-lint.yml:
```diff
@@ -0,0 +1,5 @@
+# .config/ansible-lint.yml
+warn_list:  # or 'skip_list' to silence them completely
+  - experimental  # all rules tagged as experimental
+  - latest[git]  # Result of the command may vary on subsequent runs.
```
This is idempotence check on git commands, and would fail on the presentation deployment playbook, which deploys on idr0-slot3.
Did not manage to specifically disable that check on this file.

pwalczysko commented:
This build is green now. All 9 jobs @sbesson @jburel - again, no need to reduce testing here for now. Happy to adjust to facilitate easier review.

sbesson reviewed (approved):
+1,496 −1,239 is quite a huge difference to review although from a quick code review, most of this differences comes from:

- whitespace/indentation/line-width changes
- addition of metadata e.g. `name: ...`
- usage of FQDN for consuming the Ansible modules

I did not spot anything particular concerning so happy for this to be integrated from a code review perspective. In addition, I would suggest to plan a round of functional testing i.e. select either a few representative playbooks (or all of them if capacity allows) and run them both in check-mode and outside check-mode mode against staging or production systems. 

I'll mark this as approved and let you discuss and assign reviewers for the next phase if you agree with the proposal above.

pwalczysko commented:
- [ ] training playbook check mode
- [ ] training playbook test
- [ ] demo playbook check mode
- [ ] demo playbook test
- [ ] learning check mode
- [ ] learning test
- [ ] dundeeomero check mode
- [ ] dundeeomero test
- [ ] nightshade clients check mode
- [ ] nightshade clients test
- [ ] release-www check mode
- [ ] release-www test
- [ ] web-proxy check
- [ ] web-proxy
- [ ] docker-prod check
- [ ] docker-prod

@sbesson thank you. The checklist is above. But looking at https://github.com/ome/prod-playbooks/actions/runs/5271007864 I would claim that actually all the playbooks are being run except ``demo`` and ``web-proxy`` ? Would you please advise against which servers would I run such playbook as ``web-proxy`` please ?
The test in learning and dundeeomero cases will be difficult - maybe we could do it over the summer, but not sure if the timing here is okay with that cc @jburel

sbesson commented:
You'll need to review the group these playbooks apply to and compare them to the inventory files. Some groups have staging and production hosts while others have only production hosts.

sbesson commented:
This is a whitespace diff and should be harmless so no objections from me. I'll defer to @jburel who has been working on this proxy server recently.

jburel commented:
I think it is harmless too. No objection

pwalczysko commented:
> I think it is harmless too. No objection

Thanks both. Rerun successfully without the ``--check`` flag.

pwalczysko commented:
So this is now training playbook tested in check and wet run. Note, I had to add https://github.com/ome/prod-playbooks/pull/370/commits/60682df9da31b33e66c13cf6f548f32f3d0b807a to fix a problem introduced during linting of the ``letsencrypt.yml``[1], but now it is all green again and running locally too.

pwalczysko commented:
When running the ``permissions.yml`` I have some mode changes in check mode of ansible:
Definitely did not change these modes in the playbook myself.
Okay to run in wet run cc @sbesson @jburel ?

```
ansible-playbook -i /Users/pwalczysko/Work/management_tools/ansible/inventory/prod-hosts --ask-become release/permissions.yml -l idr0-slot3.openmicroscopy.org --diff --check
BECOME password: 
[WARNING]: Invalid characters were found in group names but not replaced, use
-vvvv to see details


PLAY [Permissions] *************************************************************

TASK [Gathering Facts] *********************************************************
ok: [idr0-slot3.openmicroscopy.org]

TASK [Set perms /uod/idr/www] **************************************************
ok: [idr0-slot3.openmicroscopy.org]

TASK [Set perms2] **************************************************************
ok: [idr0-slot3.openmicroscopy.org] => (item=/uod/idr/www/docs.openmicroscopy.org)
ok: [idr0-slot3.openmicroscopy.org] => (item=/uod/idr/www/downloads.openmicroscopy.org)

TASK [Set perms3 www_folders] **************************************************
ok: [idr0-slot3.openmicroscopy.org]

TASK [Set perms3] **************************************************************
ok: [idr0-slot3.openmicroscopy.org] => (item=/uod/idr/www/docs.openmicroscopy.org/omero-romio)
ok: [idr0-slot3.openmicroscopy.org] => (item=/uod/idr/www/docs.openmicroscopy.org/ome-model)
ok: [idr0-slot3.openmicroscopy.org] => (item=/uod/idr/www/docs.openmicroscopy.org/internal)
ok: [idr0-slot3.openmicroscopy.org] => (item=/uod/idr/www/docs.openmicroscopy.org/omero-gateway)
--- before
+++ after
@@ -1,4 +1,4 @@
 {
-    "mode": "0775",
+    "mode": "01775",
     "path": "/uod/idr/www/docs.openmicroscopy.org/contributing"
 }

changed: [idr0-slot3.openmicroscopy.org] => (item=/uod/idr/www/docs.openmicroscopy.org/contributing)
ok: [idr0-slot3.openmicroscopy.org] => (item=/uod/idr/www/docs.openmicroscopy.org/bio-formats)
ok: [idr0-slot3.openmicroscopy.org] => (item=/uod/idr/www/docs.openmicroscopy.org/ome-files-cpp)
ok: [idr0-slot3.openmicroscopy.org] => (item=/uod/idr/www/docs.openmicroscopy.org/sitestyling)
ok: [idr0-slot3.openmicroscopy.org] => (item=/uod/idr/www/docs.openmicroscopy.org/omero-renderer)
ok: [idr0-slot3.openmicroscopy.org] => (item=/uod/idr/www/docs.openmicroscopy.org/omero-blitz)
ok: [idr0-slot3.openmicroscopy.org] => (item=/uod/idr/www/docs.openmicroscopy.org/omero-common)
ok: [idr0-slot3.openmicroscopy.org] => (item=/uod/idr/www/docs.openmicroscopy.org/omero-model)
ok: [idr0-slot3.openmicroscopy.org] => (item=/uod/idr/www/docs.openmicroscopy.org/contributing_backup)
ok: [idr0-slot3.openmicroscopy.org] => (item=/uod/idr/www/docs.openmicroscopy.org/ome-cmake-superbuild)
ok: [idr0-slot3.openmicroscopy.org] => (item=/uod/idr/www/docs.openmicroscopy.org/omero-server)
ok: [idr0-slot3.openmicroscopy.org] => (item=/uod/idr/www/docs.openmicroscopy.org/omero)
ok: [idr0-slot3.openmicroscopy.org] => (item=/uod/idr/www/downloads.openmicroscopy.org/logs)
ok: [idr0-slot3.openmicroscopy.org] => (item=/uod/idr/www/downloads.openmicroscopy.org/ome-model)
ok: [idr0-slot3.openmicroscopy.org] => (item=/uod/idr/www/downloads.openmicroscopy.org/omero-virtual-appliance)
ok: [idr0-slot3.openmicroscopy.org] => (item=/uod/idr/www/downloads.openmicroscopy.org/mtools)
ok: [idr0-slot3.openmicroscopy.org] => (item=/uod/idr/www/downloads.openmicroscopy.org/ome-qtwidgets)
ok: [idr0-slot3.openmicroscopy.org] => (item=/uod/idr/www/downloads.openmicroscopy.org/searcher)
ok: [idr0-slot3.openmicroscopy.org] => (item=/uod/idr/www/downloads.openmicroscopy.org/webtagging)
ok: [idr0-slot3.openmicroscopy.org] => (item=/uod/idr/www/downloads.openmicroscopy.org/idr)
ok: [idr0-slot3.openmicroscopy.org] => (item=/uod/idr/www/downloads.openmicroscopy.org/newsletters)
ok: [idr0-slot3.openmicroscopy.org] => (item=/uod/idr/www/downloads.openmicroscopy.org/bio-formats)
ok: [idr0-slot3.openmicroscopy.org] => (item=/uod/idr/www/downloads.openmicroscopy.org/movies)
ok: [idr0-slot3.openmicroscopy.org] => (item=/uod/idr/www/downloads.openmicroscopy.org/figure)
ok: [idr0-slot3.openmicroscopy.org] => (item=/uod/idr/www/downloads.openmicroscopy.org/flimfit)
ok: [idr0-slot3.openmicroscopy.org] => (item=/uod/idr/www/downloads.openmicroscopy.org/ome-files-cpp)
ok: [idr0-slot3.openmicroscopy.org] => (item=/uod/idr/www/downloads.openmicroscopy.org/ome)
--- before
+++ after
@@ -1,5 +1,5 @@
 {
-    "group": 1501,
-    "mode": "03775",
+    "group": 8000,
+    "mode": "01775",
     "path": "/uod/idr/www/downloads.openmicroscopy.org/presentations"
 }

changed: [idr0-slot3.openmicroscopy.org] => (item=/uod/idr/www/downloads.openmicroscopy.org/presentations)
ok: [idr0-slot3.openmicroscopy.org] => (item=/uod/idr/www/downloads.openmicroscopy.org/images)
ok: [idr0-slot3.openmicroscopy.org] => (item=/uod/idr/www/downloads.openmicroscopy.org/sitestyling)
ok: [idr0-slot3.openmicroscopy.org] => (item=/uod/idr/www/downloads.openmicroscopy.org/bio-formats-cpp)
ok: [idr0-slot3.openmicroscopy.org] => (item=/uod/idr/www/downloads.openmicroscopy.org/review)
ok: [idr0-slot3.openmicroscopy.org] => (item=/uod/idr/www/downloads.openmicroscopy.org/help)
ok: [idr0-slot3.openmicroscopy.org] => (item=/uod/idr/www/downloads.openmicroscopy.org/ome-common-cpp)
ok: [idr0-slot3.openmicroscopy.org] => (item=/uod/idr/www/downloads.openmicroscopy.org/resources)
ok: [idr0-slot3.openmicroscopy.org] => (item=/uod/idr/www/downloads.openmicroscopy.org/u-track)
ok: [idr0-slot3.openmicroscopy.org] => (item=/uod/idr/www/downloads.openmicroscopy.org/private)
ok: [idr0-slot3.openmicroscopy.org] => (item=/uod/idr/www/downloads.openmicroscopy.org/editor)
ok: [idr0-slot3.openmicroscopy.org] => (item=/uod/idr/www/downloads.openmicroscopy.org/patches)
ok: [idr0-slot3.openmicroscopy.org] => (item=/uod/idr/www/downloads.openmicroscopy.org/ome-cmake-superbuild)
ok: [idr0-slot3.openmicroscopy.org] => (item=/uod/idr/www/downloads.openmicroscopy.org/ice)
ok: [idr0-slot3.openmicroscopy.org] => (item=/uod/idr/www/downloads.openmicroscopy.org/web-archive)
ok: [idr0-slot3.openmicroscopy.org] => (item=/uod/idr/www/downloads.openmicroscopy.org/omero)

PLAY RECAP *********************************************************************
idr0-slot3.openmicroscopy.org : ok=5    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
```

pwalczysko commented:
again, some changes highlighted when the ``public-images`` playbook is run

```
ansible-playbook -i /Users/pwalczysko/Work/management_tools/ansible/inventory/prod-hosts --ask-become release/public-images.yml -l idr0-slot3.openmicroscopy.org --diff --check 
BECOME password: 
[WARNING]: Invalid characters were found in group names but not replaced, use
-vvvv to see details


PLAY [Public images] ***********************************************************

TASK [Gathering Facts] *********************************************************
ok: [idr0-slot3.openmicroscopy.org]

TASK [Create symlinks for public images] ***************************************
--- before
+++ after
@@ -1,4 +1,4 @@
 {
     "path": "\"/uod/idr/www/downloads. openmicroscopy.org/ images/AmiraMesh\"\n",
-    "state": "absent"
+    "state": "link"
 }

changed: [idr0-slot3.openmicroscopy.org] => (item={'key': 'amira', 'value': 'AmiraMesh'})
--- before
+++ after
@@ -1,4 +1,4 @@
 {
     "path": "\"/uod/idr/www/downloads. openmicroscopy.org/ images/BDV\"\n",
-    "state": "absent"
+    "state": "link"
 }

changed: [idr0-slot3.openmicroscopy.org] => (item={'key': 'bdv', 'value': 'BDV'})
--- before
+++ after
@@ -1,4 +1,4 @@
 {
     "path": "\"/uod/idr/www/downloads. openmicroscopy.org/ images/SPC-FIFO\"\n",
-    "state": "absent"
+    "state": "link"
 }

changed: [idr0-slot3.openmicroscopy.org] => (item={'key': 'becker-hickl-spc', 'value': 'SPC-FIFO'})
--- before
+++ after
@@ -1,4 +1,4 @@
 {
     "path": "\"/uod/idr/www/downloads. openmicroscopy.org/ images/Cellomics\"\n",
-    "state": "absent"
+    "state": "link"
 }

changed: [idr0-slot3.openmicroscopy.org] => (item={'key': 'cellomics', 'value': 'Cellomics'})
--- before
+++ after
@@ -1,4 +1,4 @@
 {
     "path": "\"/uod/idr/www/downloads. openmicroscopy.org/ images/CellH5\"\n",
-    "state": "absent"
+    "state": "link"
 }

changed: [idr0-slot3.openmicroscopy.org] => (item={'key': 'cellh5', 'value': 'CellH5'})
--- before
+++ after
@@ -1,4 +1,4 @@
 {
     "path": "\"/uod/idr/www/downloads. openmicroscopy.org/ images/CellWorX\"\n",
-    "state": "absent"
+    "state": "link"
 }

changed: [idr0-slot3.openmicroscopy.org] => (item={'key': 'cellworx', 'value': 'CellWorX'})
--- before
+++ after
@@ -1,4 +1,4 @@
 {
     "path": "\"/uod/idr/www/downloads. openmicroscopy.org/ images/PerkinElmer-Columbus\"\n",
-    "state": "absent"
+    "state": "link"
 }

changed: [idr0-slot3.openmicroscopy.org] => (item={'key': 'columbus', 'value': 'PerkinElmer-Columbus'})
--- before
+++ after
@@ -1,4 +1,4 @@
 {
     "path": "\"/uod/idr/www/downloads. openmicroscopy.org/ images/DV\"\n",
-    "state": "absent"
+    "state": "link"
 }

changed: [idr0-slot3.openmicroscopy.org] => (item={'key': 'deltavision', 'value': 'DV'})
--- before
+++ after
@@ -1,4 +1,4 @@
 {
     "path": "\"/uod/idr/www/downloads. openmicroscopy.org/ images/DICOM\"\n",
-    "state": "absent"
+    "state": "link"
 }

changed: [idr0-slot3.openmicroscopy.org] => (item={'key': 'dicom', 'value': 'DICOM'})
--- before
+++ after
@@ -1,4 +1,4 @@
 {
     "path": "\"/uod/idr/www/downloads. openmicroscopy.org/ images/ECAT7\"\n",
-    "state": "absent"
+    "state": "link"
 }

changed: [idr0-slot3.openmicroscopy.org] => (item={'key': 'ecat7', 'value': 'ECAT7'})
--- before
+++ after
@@ -1,4 +1,4 @@
 {
     "path": "\"/uod/idr/www/downloads. openmicroscopy.org/ images/Flex\"\n",
-    "state": "absent"
+    "state": "link"
 }

changed: [idr0-slot3.openmicroscopy.org] => (item={'key': 'flex', 'value': 'Flex'})
--- before
+++ after
@@ -1,4 +1,4 @@
 {
     "path": "\"/uod/idr/www/downloads. openmicroscopy.org/ images/Olympus-FluoView\"\n",
-    "state": "absent"
+    "state": "link"
 }

changed: [idr0-slot3.openmicroscopy.org] => (item={'key': 'fv1000', 'value': 'Olympus-FluoView'})
--- before
+++ after
@@ -1,4 +1,4 @@
 {
     "path": "\"/uod/idr/www/downloads. openmicroscopy.org/ images/Gatan\"\n",
-    "state": "absent"
+    "state": "link"
 }

changed: [idr0-slot3.openmicroscopy.org] => (item={'key': 'gatan', 'value': 'Gatan'})
--- before
+++ after
@@ -1,4 +1,4 @@
 {
     "path": "\"/uod/idr/www/downloads. openmicroscopy.org/ images/Hamamatsu-NDPI\"\n",
-    "state": "absent"
+    "state": "link"
 }

changed: [idr0-slot3.openmicroscopy.org] => (item={'key': 'hamamatsu', 'value': 'Hamamatsu-NDPI'})
--- before
+++ after
@@ -1,4 +1,4 @@
 {
     "path": "\"/uod/idr/www/downloads. openmicroscopy.org/ images/Hamamatsu-VMS\"\n",
-    "state": "absent"
+    "state": "link"
 }

changed: [idr0-slot3.openmicroscopy.org] => (item={'key': 'hamamatsu-vms', 'value': 'Hamamatsu-VMS'})
--- before
+++ after
@@ -1,4 +1,4 @@
 {
     "path": "\"/uod/idr/www/downloads. openmicroscopy.org/ images/ICS\"\n",
-    "state": "absent"
+    "state": "link"
 }

changed: [idr0-slot3.openmicroscopy.org] => (item={'key': 'ics', 'value': 'ICS'})
--- before
+++ after
@@ -1,4 +1,4 @@
 {
     "path": "\"/uod/idr/www/downloads. openmicroscopy.org/ images/InCell3000\"\n",
-    "state": "absent"
+    "state": "link"
 }

changed: [idr0-slot3.openmicroscopy.org] => (item={'key': 'incell3000', 'value': 'InCell3000'})
--- before
+++ after
@@ -1,4 +1,4 @@
 {
     "path": "\"/uod/idr/www/downloads. openmicroscopy.org/ images/InCell2000\"\n",
-    "state": "absent"
+    "state": "link"
 }

changed: [idr0-slot3.openmicroscopy.org] => (item={'key': 'incell', 'value': 'InCell2000'})
--- before
+++ after
@@ -1,4 +1,4 @@
 {
     "path": "\"/uod/idr/www/downloads. openmicroscopy.org/ images/Imaris-IMS\"\n",
-    "state": "absent"
+    "state": "link"
 }

changed: [idr0-slot3.openmicroscopy.org] => (item={'key': 'imaris', 'value': 'Imaris-IMS'})
--- before
+++ after
@@ -1,4 +1,4 @@
 {
     "path": "\"/uod/idr/www/downloads. openmicroscopy.org/ images/KLB\"\n",
-    "state": "absent"
+    "state": "link"
 }

changed: [idr0-slot3.openmicroscopy.org] => (item={'key': 'klb', 'value': 'KLB'})
--- before
+++ after
@@ -1,4 +1,4 @@
 {
     "path": "\"/uod/idr/www/downloads. openmicroscopy.org/ images/Leica-LIF\"\n",
-    "state": "absent"
+    "state": "link"
 }

changed: [idr0-slot3.openmicroscopy.org] => (item={'key': 'leica-lif', 'value': 'Leica-LIF'})
--- before
+++ after
@@ -1,4 +1,4 @@
 {
     "path": "\"/uod/idr/www/downloads. openmicroscopy.org/ images/Leica-SCN\"\n",
-    "state": "absent"
+    "state": "link"
 }

changed: [idr0-slot3.openmicroscopy.org] => (item={'key': 'leica-scn', 'value': 'Leica-SCN'})
--- before
+++ after
@@ -1,4 +1,4 @@
 {
     "path": "\"/uod/idr/www/downloads. openmicroscopy.org/ images/Leica-XLEF\"\n",
-    "state": "absent"
+    "state": "link"
 }

changed: [idr0-slot3.openmicroscopy.org] => (item={'key': 'leica-xlef-lof', 'value': 'Leica-XLEF'})
--- before
+++ after
@@ -1,4 +1,4 @@
 {
     "path": "\"/uod/idr/www/downloads. openmicroscopy.org/ images/LEO\"\n",
-    "state": "absent"
+    "state": "link"
 }

changed: [idr0-slot3.openmicroscopy.org] => (item={'key': 'leo', 'value': 'LEO'})
--- before
+++ after
@@ -1,4 +1,4 @@
 {
     "path": "\"/uod/idr/www/downloads. openmicroscopy.org/ images/MetaXpress\"\n",
-    "state": "absent"
+    "state": "link"
 }

changed: [idr0-slot3.openmicroscopy.org] => (item={'key': 'metaxpress', 'value': 'MetaXpress'})
--- before
+++ after
@@ -1,4 +1,4 @@
 {
     "path": "\"/uod/idr/www/downloads. openmicroscopy.org/ images/Micro-Manager\"\n",
-    "state": "absent"
+    "state": "link"
 }

changed: [idr0-slot3.openmicroscopy.org] => (item={'key': 'micromanager', 'value': 'Micro-Manager'})
--- before
+++ after
@@ -1,4 +1,4 @@
 {
     "path": "\"/uod/idr/www/downloads. openmicroscopy.org/ images/MRC\"\n",
-    "state": "absent"
+    "state": "link"
 }

changed: [idr0-slot3.openmicroscopy.org] => (item={'key': 'mrc', 'value': 'MRC'})
--- before
+++ after
@@ -1,4 +1,4 @@
 {
     "path": "\"/uod/idr/www/downloads. openmicroscopy.org/ images/ND2\"\n",
-    "state": "absent"
+    "state": "link"
 }

changed: [idr0-slot3.openmicroscopy.org] => (item={'key': 'nd2', 'value': 'ND2'})
--- before
+++ after
@@ -1,4 +1,4 @@
 {
     "path": "\"/uod/idr/www/downloads. openmicroscopy.org/ images/NIfTI\"\n",
-    "state": "absent"
+    "state": "link"
 }

changed: [idr0-slot3.openmicroscopy.org] => (item={'key': 'nifti', 'value': 'NIfTI'})
--- before
+++ after
@@ -1,4 +1,4 @@
 {
     "path": "\"/uod/idr/www/downloads. openmicroscopy.org/ images/NRRD\"\n",
-    "state": "absent"
+    "state": "link"
 }

changed: [idr0-slot3.openmicroscopy.org] => (item={'key': 'nrrd', 'value': 'NRRD'})
--- before
+++ after
@@ -1,4 +1,4 @@
 {
     "path": "\"/uod/idr/www/downloads. openmicroscopy.org/ images/OBF\"\n",
-    "state": "absent"
+    "state": "link"
 }

changed: [idr0-slot3.openmicroscopy.org] => (item={'key': 'obf', 'value': 'OBF'})
--- before
+++ after
@@ -1,4 +1,4 @@
 {
     "path": "\"/uod/idr/www/downloads. openmicroscopy.org/ images/Olympus-OIR\"\n",
-    "state": "absent"
+    "state": "link"
 }

changed: [idr0-slot3.openmicroscopy.org] => (item={'key': 'olympus-oir', 'value': 'Olympus-OIR'})
--- before
+++ after
@@ -1,4 +1,4 @@
 {
     "path": "\"/uod/idr/www/downloads. openmicroscopy.org/ images/OME-XML\"\n",
-    "state": "absent"
+    "state": "link"
 }

changed: [idr0-slot3.openmicroscopy.org] => (item={'key': 'ome-xml', 'value': 'OME-XML'})
--- before
+++ after
@@ -1,4 +1,4 @@
 {
     "path": "\"/uod/idr/www/downloads. openmicroscopy.org/ images/OME-TIFF\"\n",
-    "state": "absent"
+    "state": "link"
 }

changed: [idr0-slot3.openmicroscopy.org] => (item={'key': 'ome-tiff', 'value': 'OME-TIFF'})
--- before
+++ after
@@ -1,4 +1,4 @@
 {
     "path": "\"/uod/idr/www/downloads. openmicroscopy.org/ images/PerkinElmer-Operetta\"\n",
-    "state": "absent"
+    "state": "link"
 }

changed: [idr0-slot3.openmicroscopy.org] => (item={'key': 'perkinelmer-operetta', 'value': 'PerkinElmer-Operetta'})
--- before
+++ after
@@ -1,4 +1,4 @@
 {
     "path": "\"/uod/idr/www/downloads. openmicroscopy.org/ images/PNG\"\n",
-    "state": "absent"
+    "state": "link"
 }

changed: [idr0-slot3.openmicroscopy.org] => (item={'key': 'png', 'value': 'PNG'})
--- before
+++ after
@@ -1,4 +1,4 @@
 {
     "path": "\"/uod/idr/www/downloads. openmicroscopy.org/ images/SVS\"\n",
-    "state": "absent"
+    "state": "link"
 }

changed: [idr0-slot3.openmicroscopy.org] => (item={'key': 'svs', 'value': 'SVS'})
--- before
+++ after
@@ -1,4 +1,4 @@
 {
     "path": "\"/uod/idr/www/downloads. openmicroscopy.org/ images/TIFF\"\n",
-    "state": "absent"
+    "state": "link"
 }

changed: [idr0-slot3.openmicroscopy.org] => (item={'key': 'tiff', 'value': 'TIFF'})
--- before
+++ after
@@ -1,4 +1,4 @@
 {
     "path": "\"/uod/idr/www/downloads. openmicroscopy.org/ images/Trestle\"\n",
-    "state": "absent"
+    "state": "link"
 }

changed: [idr0-slot3.openmicroscopy.org] => (item={'key': 'trestle', 'value': 'Trestle'})
--- before
+++ after
@@ -1,4 +1,4 @@
 {
     "path": "\"/uod/idr/www/downloads. openmicroscopy.org/ images/Ventana\"\n",
-    "state": "absent"
+    "state": "link"
 }

changed: [idr0-slot3.openmicroscopy.org] => (item={'key': 'ventana', 'value': 'Ventana'})
--- before
+++ after
@@ -1,4 +1,4 @@
 {
     "path": "\"/uod/idr/www/downloads. openmicroscopy.org/ images/Vectra-QPTIFF\"\n",
-    "state": "absent"
+    "state": "link"
 }

changed: [idr0-slot3.openmicroscopy.org] => (item={'key': 'vectra-qptiff', 'value': 'Vectra-QPTIFF'})
--- before
+++ after
@@ -1,4 +1,4 @@
 {
     "path": "\"/uod/idr/www/downloads. openmicroscopy.org/ images/Zeiss-CZI\"\n",
-    "state": "absent"
+    "state": "link"
 }

changed: [idr0-slot3.openmicroscopy.org] => (item={'key': 'zeiss-czi', 'value': 'Zeiss-CZI'})

TASK [Check public images] *****************************************************
--- before
+++ after
@@ -1,4 +1,4 @@
 {
     "path": "\"/uod/idr/www/downloads. openmicroscopy.org/ images/gateway_tests/big.tiff\"\n",
-    "state": "absent"
+    "state": "link"
 }

changed: [idr0-slot3.openmicroscopy.org] => (item={'src': '../../../../repos/curated/samples/carlos/big.tiff', 'dest': 'gateway_tests/big.tiff'})
--- before
+++ after
@@ -1,4 +1,4 @@
 {
     "path": "\"/uod/idr/www/downloads. openmicroscopy.org/ images/gateway_tests/CHOBI_d3d.dv\"\n",
-    "state": "absent"
+    "state": "link"
 }

changed: [idr0-slot3.openmicroscopy.org] => (item={'src': '../../../../repos/curated/samples/ome/CHOBI_d3d.dv', 'dest': 'gateway_tests/CHOBI_d3d.dv'})
--- before
+++ after
@@ -1,4 +1,4 @@
 {
     "path": "\"/uod/idr/www/downloads. openmicroscopy.org/ images/gateway_tests/tinyTest.d3d.dv\"\n",
-    "state": "absent"
+    "state": "link"
 }

changed: [idr0-slot3.openmicroscopy.org] => (item={'src': '../../../../repos/curated/samples/ome/tinyTest.d3d.dv', 'dest': 'gateway_tests/tinyTest.d3d.dv'})
--- before
+++ after
@@ -1,4 +1,4 @@
 {
     "path": "\"/uod/idr/www/downloads. openmicroscopy.org/ images/HCS/BBBC\"\n",
-    "state": "absent"
+    "state": "link"
 }

changed: [idr0-slot3.openmicroscopy.org] => (item={'src': '../../../../repos/curated/cellomics/public/', 'dest': 'HCS/BBBC'})
--- before
+++ after
@@ -1,4 +1,4 @@
 {
     "path": "\"/uod/idr/www/downloads. openmicroscopy.org/ images/HCS/INCELL2000\"\n",
-    "state": "absent"
+    "state": "link"
 }

changed: [idr0-slot3.openmicroscopy.org] => (item={'src': '../../../../repos/curated/incell/public/', 'dest': 'HCS/INCELL2000'})
--- before
+++ after
@@ -1,4 +1,4 @@
 {
     "path": "\"/uod/idr/www/downloads. openmicroscopy.org/ images/HCS/Operetta\"\n",
-    "state": "absent"
+    "state": "link"
 }

changed: [idr0-slot3.openmicroscopy.org] => (item={'src': '../../../../repos/curated/perkinelmer-operetta/public/', 'dest': 'HCS/Operetta'})
--- before
+++ after
@@ -1,4 +1,4 @@
 {
     "path": "\"/uod/idr/www/downloads. openmicroscopy.org/ images/u-track\"\n",
-    "state": "absent"
+    "state": "link"
 }

changed: [idr0-slot3.openmicroscopy.org] => (item={'src': '../../../repos/curated/zip/u-track/', 'dest': 'u-track'})

PLAY RECAP *********************************************************************
idr0-slot3.openmicroscopy.org : ok=3    changed=2    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0 
```

pwalczysko commented:
The release acceptance playbook: The way of how to run it is not clear - the ``version`` and ``product`` variables seem to be hard to get right.

```
ansible-playbook -i /Users/pwalczysko/Work/management_tools/ansible/inventory/prod-hosts --ask-become release/release-acceptance.yml -e version=latest -e product=latest  -l idr0-slot3.openmicroscopy.org --diff --check
BECOME password: 
[WARNING]: Invalid characters were found in group names but not replaced, use
-vvvv to see details


PLAY [Release acceptance] ******************************************************

TASK [Gathering Facts] *********************************************************
ok: [idr0-slot3.openmicroscopy.org]

TASK [Check mandatory variables are defined] ***********************************
skipping: [idr0-slot3.openmicroscopy.org]

TASK [Check the release component exist] ***************************************
ok: [idr0-slot3.openmicroscopy.org] => (item=/uod/idr/www/docs.openmicroscopy.org)
ok: [idr0-slot3.openmicroscopy.org] => (item=/uod/idr/www/downloads.openmicroscopy.org)

TASK [Check the release component exist] ***************************************
failed: [idr0-slot3.openmicroscopy.org] (item={'invocation': {'module_args': {'checksum_algorithm': 'sha1', 'get_checksum': True, 'follow': False, 'path': '/uod/idr/www/docs.openmicroscopy.org/latest/latest/', 'get_md5': None, 'get_mime': True, 'get_attributes': True}}, 'stat': {'exists': False}, 'changed': False, 'failed': False, 'item': '/uod/idr/www/docs.openmicroscopy.org', 'ansible_loop_var': 'item'}) => {"ansible_loop_var": "item", "changed": false, "item": {"ansible_loop_var": "item", "changed": false, "failed": false, "invocation": {"module_args": {"checksum_algorithm": "sha1", "follow": false, "get_attributes": true, "get_checksum": true, "get_md5": null, "get_mime": true, "path": "/uod/idr/www/docs.openmicroscopy.org/latest/latest/"}}, "item": "/uod/idr/www/docs.openmicroscopy.org", "stat": {"exists": false}}, "module_stderr": "Shared connection to idr0-slot3.openmicroscopy.org closed.\r\n", "module_stdout": "\r\n", "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error", "rc": 0}
failed: [idr0-slot3.openmicroscopy.org] (item={'invocation': {'module_args': {'checksum_algorithm': 'sha1', 'get_checksum': True, 'follow': False, 'path': '/uod/idr/www/downloads.openmicroscopy.org/latest/latest/', 'get_md5': None, 'get_mime': True, 'get_attributes': True}}, 'stat': {'exists': False}, 'changed': False, 'failed': False, 'item': '/uod/idr/www/downloads.openmicroscopy.org', 'ansible_loop_var': 'item'}) => {"ansible_loop_var": "item", "changed": false, "item": {"ansible_loop_var": "item", "changed": false, "failed": false, "invocation": {"module_args": {"checksum_algorithm": "sha1", "follow": false, "get_attributes": true, "get_checksum": true, "get_md5": null, "get_mime": true, "path": "/uod/idr/www/downloads.openmicroscopy.org/latest/latest/"}}, "item": "/uod/idr/www/downloads.openmicroscopy.org", "stat": {"exists": false}}, "module_stderr": "Shared connection to idr0-slot3.openmicroscopy.org closed.\r\n", "module_stdout": "\r\n", "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error", "rc": 0}

PLAY RECAP *********************************************************************
idr0-slot3.openmicroscopy.org : ok=2    changed=0    unreachable=0    failed=1    skipped=1    rescued=0    ignored=0   

```

sbesson commented:
You might want to use a real example e.g. `product=omero version=5.6.7` although I am not sure how much the `release-acceptance` playbook is WIP with the ongoing readthedocs migration. @jburel will be the authority here

pwalczysko commented:
> You might want to use a real example e.g. `product=omero version=5.6.7` although I am not sure how much the `release-acceptance` playbook is WIP with the ongoing readthedocs migration. @jburel will be the authority here

Thank you @sbesson , that fixed the run of the ``release-acceptance`` - main thing it passes :)

pwalczysko commented:
Repeating the checklist here as it is finished now fmpov:

Passing tests (always the check mode was run, the test mode, ie. wet run was run only when absolutely safe).
- [x] training playbook check mode
- [x] training playbook test
- [x] demo playbook check mode
- [x] demo playbook test
- [x] learning check mode
- [ ] learning test
- [x] dundeeomero check mode
- [ ] dundeeomero test
- [x] nightshade clients check mode
- [ ] nightshade clients test
- [x] bootstrap/playbook check
- [ ] bootstrap/playbook test
- [x] k8s/bootstrap/playbook check
- [ ] k8s/bootstrap/playbook test
- [x] k8s/prerequisites/playbook check
- [ ] k8s/prerequisites/playbook test
- [x] omero/monitoring-agents check
- [ ] omero/monitoring-agents test
- [x] omero/firewall check
- [x] omero/firewall test
- [x] omero/sls-gallery.yml check
- [ ] omero/sls-gallery.yml test
- [x] release-perms check mode
- [ ] release-perms test
- [x] release-presentations check mode
- [x] release-presentations test
- [x] release-public-images check mode
- [ ] release-public-images test
- [x] release-acceptance check mode
- [ ] release-acceptance test
- [x] web-proxy check
- [x] web-proxy
- [x] omedev/playbook.yml check
- [ ] omedev/playbook.yml test
- [x] www check
- [x] www test
- [x] www/www-static check
- [x] www/www-static test

Failing - note that none of the failing playbooks below were changed in this PR
- [ ] maintenance/omero-restore-db.yml check (failing, but failing also on master with the same error, not changed in the PR)
- [ ] maintenance/omero-restore-db.yml test
- [ ] docker-prod-apps.yml check (playbook failing, but not changed in the PR)
- [ ] docker-prod-apps.yml test
- [ ] postgres/ome-pg-prod.yml check (playbook failing, but not changed in the PR)
- [ ] postgres/ome-pg-prod.yml test

pwalczysko commented:
Considering that the playbook runs were successful https://github.com/ome/prod-playbooks/pull/370#issuecomment-1640435252 for all the playbooks changed in this PR at least in check mode, and there was no indication that the wet run of those would uncover new errors, I would claim that the testing of the PR is sufficient. The failing playbooks from https://github.com/ome/prod-playbooks/pull/370#issuecomment-1640435252 were not changed in the PR, I am listing them just for the completeness sake.

cc @sbesson @jburel

pwalczysko commented:
@jburel Happy to merge ?
